server/plugin.go

```diff
@@ -599,22 +670,84 @@ func (p *Plugin) HasGroupHook(ctx context.Context, user *gitlab.UserInfo, namesp
 	return found, err
 }
 
-func (p *Plugin) checkAndRefreshToken(token *oauth2.Token) (*oauth2.Token, error) {
-	// If there is only one minute left for the token to expire, we are refreshing the token.
-	// The detailed reason for this can be found here: https://github.com/golang/oauth2/issues/84#issuecomment-831492464
-	// We don't want the token to expire between the time when we decide that the old token is valid
-	// and the time at which we create the request. We are handling that by not letting the token expire.
-	if time.Until(token.Expiry) <= 1*time.Minute {
-		conf := p.getOAuthConfig()
-		src := conf.TokenSource(context.Background(), token)
-		newToken, err := src.Token() // this actually goes and renews the tokens
-		if err != nil {
-			return nil, errors.Wrap(err, "unable to get the new refreshed token")
+func (p *Plugin) refreshToken(userInfo *gitlab.UserInfo, token *oauth2.Token) (*oauth2.Token, error) {
+	conf := p.getOAuthConfig()
+	src := conf.TokenSource(context.Background(), token)
+
+	newToken, err := src.Token() // this actually goes and renews the tokens
+
+	if err != nil {
+		if strings.Contains(err.Error(), "\"error\":\"invalid_grant\"") {
+			p.handleRevokedToken(userInfo)
```

Comment:
Just so I understand, this will only run when we have a token stored, and the token was revoked externally. In this case, the mutex will correctly only allow this to run once.

Reply:
This will only run when the token was revoked externally, and the token expiration time is within one minute of the request (since this is in `p.refreshToken`, it occurs when we try to refresh).

If the token is not marked to expire, we'll try to use it and it will fail with the `401 {error: invalid_token}` error (a different error when we try to make a request through the `go-gitlab` library). And if it fails there, we call `p.handleRevokedToken` from `useGitlabClient` so the result is the same.